core/src/main/java/org/stellar/anchor/sep6/ExchangeAmountsCalculator.java

```diff
@@ -124,4 +132,42 @@ public static class Amounts {
     String amountFee;
     String amountFeeAsset;
   }
+
+  private String getClientId(Sep10Jwt token) throws BadRequestException {
+    ClientsConfig.ClientConfig clientByDomain =
+        clientsConfig.getClientConfigByDomain(token.getClientDomain());
+    ClientsConfig.ClientConfig clientByAccount =
+        clientsConfig.getClientConfigBySigningKey(token.getAccount());
+    ClientsConfig.ClientConfig client = clientByDomain != null ? clientByDomain : clientByAccount;
+
+    if (!sep10Config.isClientAttributionRequired()) {
+      return client != null ? client.getName() : null;
+    }
+    if (sep10Config.isClientAttributionRequired() && client == null) {
+      throw new BadRequestException("Client not found");
+    }
+
+    if (sep10Config.getAllowedClientDomains().isEmpty()
+        && sep10Config.getAllowedClientNames().isEmpty()) {
+      return client.getName();
+    }
+
+    if (token.getClientDomain() != null) {
+      if (sep10Config.getAllowedClientDomains().contains(client.getDomain())
+          || sep10Config.getAllowedClientDomains().isEmpty()) {
+        if (sep10Config.getAllowedClientNames().contains(client.getName())
+            || sep10Config.getAllowedClientNames().isEmpty()) {
+          return client.getName();
+        }
+        throw new BadRequestException("Client name not allowed");
+      }
+      throw new BadRequestException("Client domain not allowed");
+    }
+
+    if (sep10Config.getAllowedClientNames().contains(client.getName())
+        || sep10Config.getAllowedClientNames().isEmpty()) {
+      return client.getName();
+    }
+    throw new BadRequestException("Client name not allowed");
```

Comment:
The logic seems complicated and not documented. Can we add some reference?

Reply:
Sure. I'll add a Javadoc and annotate the sections.

Reply:
would it make sense to actually break this into smaller methods? would be easeier to read and be more testable?